Add remaining support for the browser.scripting API.

kiaraarose opened the pull request:
#### 8f5f542c3b969fc0e112188544b4ddefa0b8a182
<pre>
Add remaining support for the browser.scripting API.
<a href="https://bugs.webkit.org/show_bug.cgi?id=259956">https://bugs.webkit.org/show_bug.cgi?id=259956</a>

Reviewed by Timothy Hatcher.

This patch moves the code used for tabs.executeScript into the WebExtensionDynamicScripts so that
the code can be shared with the scripting.executeScript API.

Also adds tests for executeScript(), insertCSS(), and removeCSS().

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
(WebKit::WebExtensionContext::scriptingExecuteScript):
(WebKit::WebExtensionContext::scriptingInsertCSS):
(WebKit::WebExtensionContext::scriptingRemoveCSS):
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPITabsCocoa.mm:
(WebKit::WebExtensionContext::tabsExecuteScript):
(WebKit::WebExtensionContext::tabsInsertCSS):
(WebKit::WebExtensionContext::tabsRemoveCSS):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
(WebKit::WebExtensionDynamicScripts::getSourcePairsForParameters):
(WebKit::WebExtensionDynamicScripts::executeScript):
(WebKit::WebExtensionDynamicScripts::injectStyleSheets):
(WebKit::WebExtensionDynamicScripts::removeStyleSheets):
(WebKit::WebExtensionDynamicScripts::getSourcePairsForResource): Deleted.
* Source/WebKit/UIProcess/Extensions/WebExtensionDynamicScripts.h:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
(WebKit::parseTargetInjectionOptions):
(WebKit::WebExtensionAPIScripting::executeScript):
(WebKit::WebExtensionAPIScripting::toWebAPI):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIScripting.h:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/xcshareddata/xcschemes/TestWebKitAPI.xcscheme:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIScripting.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/270576@main">https://commits.webkit.org/270576@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e52f4137944d60dc276909c4248e413a78786504

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25834 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4441 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27111 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27931 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23650 "Built successfully") 
| [❌ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/26148 "Failed to checkout and rebase branch from PR 20324") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6199 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/51/builds/1868 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/23746 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26083 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/3330 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/22245 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28511 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/2948 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/23201 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/29275 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/23563 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/23576 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/27148 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2979 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/51/builds/1868 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4370 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3435 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3309 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/3295 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -134,8 +135,8 @@
 
-    Vector<std::optional<SourcePair>> styleSheetPairs = getSourcePairsForResource(parameters.files, parameters.css, m_extension);
-    removeStyleSheets(styleSheetPairs, injectedFrames, *this);
+    Vector<std::optional<SourcePair>> styleSheetPairs = getSourcePairsForParameters(parameters, m_extension);
```
`auto styleSheetPairs = …`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -100,7 +101,7 @@
     auto injectedFrames = parameters.frameIDs ? WebCore::UserContentInjectedFrames::InjectInTopFrameOnly : WebCore::UserContentInjectedFrames::InjectInAllFrames;
 
-    SourcePairs styleSheetPairs = getSourcePairsForResource(parameters.files, parameters.css, m_extension);
+    SourcePairs styleSheetPairs = getSourcePairsForParameters(parameters, m_extension);
```
`auto styleSheetPairs = …`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -73,8 +72,10 @@
 
-    // FIXME: <https://webkit.org/b/259954> Implement scripting.executeScript().
-    completionHandler({ }, std::nullopt);
+    SourcePairs scriptPairs = getSourcePairsForParameters(parameters, m_extension);
```
`auto scriptPairs = …`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPITabsCocoa.mm:
```diff
@@ -582,30 +577,10 @@ static inline String toMIMEType(WebExtensionTab::ImageFormat format)
-    auto injectionResults = InjectionResultHolder::create();
-    auto aggregator = MainRunLoopCallbackAggregator::create([injectionResults, completionHandler = WTFMove(completionHandler)]() mutable {
-        completionHandler(injectionResults->results, std::nullopt);
+    SourcePairs scriptPairs = getSourcePairsForParameters(parameters, m_extension);
```
`auto scriptPairs = …`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPITabsCocoa.mm:
```diff
@@ -630,7 +605,7 @@ static inline String toMIMEType(WebExtensionTab::ImageFormat format)
     auto injectedFrames = parameters.frameIDs ? WebCore::UserContentInjectedFrames::InjectInTopFrameOnly : WebCore::UserContentInjectedFrames::InjectInAllFrames;
 
-    SourcePairs styleSheetPairs = getSourcePairsForResource(parameters.files, parameters.code, m_extension);
+    SourcePairs styleSheetPairs = getSourcePairsForParameters(parameters, m_extension);
```
`auto styleSheetPairs = …`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPITabsCocoa.mm:
```diff
@@ -658,8 +633,8 @@ static inline String toMIMEType(WebExtensionTab::ImageFormat format)
 
-    SourcePairs styleSheetPairs = getSourcePairsForResource(parameters.files, parameters.code, m_extension);
-    removeStyleSheets(styleSheetPairs, injectedFrames, *this);
+    SourcePairs styleSheetPairs = getSourcePairsForParameters(parameters, m_extension);
```
`auto styleSheetPairs = …`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
```diff
@@ -86,36 +90,76 @@ static bool userStyleSheetMatchesContent(Ref<API::UserStyleSheet> userStyleSheet
+        completionHandler(injectionResults);
+    });
+
+    [webView _frames:makeBlockPtr([webView = RetainPtr { webView }, tab, context = Ref { context }, scriptPairs, executionWorld = Ref { executionWorld }, injectionResults, aggregator, parameters](_WKFrameTreeNode *mainFrame) mutable {
```
`tab` should be `tab = RefPtr { tab }`. And `scriptPairs = WTFMove(scriptPairs)`, same for `parameters` do `parameters  = WTFMove(parameters)`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
```diff
@@ -86,36 +90,76 @@ static bool userStyleSheetMatchesContent(Ref<API::UserStyleSheet> userStyleSheet
+                NSString *javaScript = [NSString stringWithFormat:@"return (%@)(...arguments)", (NSString *)parameters.function.value()];
+                NSArray *arguments = parameters.arguments ? createNSArray(parameters.arguments.value()).get() : @[ ];
+
+                [webView _callAsyncJavaScript:javaScript arguments:@{ @"arguments": arguments } inFrame:frameInfo inContentWorld:world completionHandler:makeBlockPtr([injectionResults, aggregator, frameInfo](id resultOfExecution, NSError *error) mutable {
```
`frameInfo` might need captured as `frameInfo = RetainPtr { frameInfo }`. Not sure how ARC is at play here…

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
```diff
@@ -86,36 +90,76 @@ static bool userStyleSheetMatchesContent(Ref<API::UserStyleSheet> userStyleSheet
+            }
+
+            for (auto& script : scriptPairs.value()) {
+                [webView _evaluateJavaScript:script.value().first withSourceURL:script.value().second.value_or(URL { }) inFrame:frameInfo inContentWorld:executionWorld->wrapper() completionHandler:makeBlockPtr([injectionResults, aggregator, frameInfo](id resultOfExecution, NSError *error) mutable {
```
Ditto for capturing `frameInfo`.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -147,7 +149,12 @@ static inline void parseCSSInjectionOptions(NSDictionary *cssInfo, WebExtensionS
     parseScriptInjectionOptions(script, parameters);
 
-    // FIXME: <https://webkit.org/b/259954> Handle script injections in the UIProcess.
+    WebProcess::singleton().sendWithAsyncReply(Messages::WebExtensionContext::ScriptingExecuteScript(WTFMove(parameters)), [protectedThis = Ref { *this }, callback = WTFMove(callback), this](std::optional<Vector<WebKit::WebExtensionScriptInjectionResultParameters>> results, WebExtensionDynamicScripts::Error error) {
```
Is `, this` needed? I think it will complain since it isn't used from what I can tell.

kiaraarose replied:
i needed it for `toWebAPI`. i got a `'this' cannot be implicitly captured in this context` without it

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -342,6 +349,31 @@ static inline void parseCSSInjectionOptions(NSDictionary *cssInfo, WebExtensionS
+        NSMutableDictionary *result = [NSMutableDictionary dictionaryWithCapacity:3];
+
+        if (parameters.result)
+            result[@"result"] = parameters.result.value();
```
This might need a `(NSString *)` cast? Maybe it is automatically picking that?

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -342,6 +349,31 @@ static inline void parseCSSInjectionOptions(NSDictionary *cssInfo, WebExtensionS
+            result[@"frameId"] = @(WebKit::toWebAPI(parameters.frameID.value()));
+
+        if (parameters.error)
+            result[@"error"] = parameters.error.value();
```
Ditto on the cast?

kiaraarose replied:
I thought you could go from String to NSString * without a cast? but i'll add

xeenon commented on Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/xcshareddata/xcschemes/TestWebKitAPI.xcscheme:
```diff
@@ -64,6 +64,12 @@
+            argument = "--gtest_filter=WKWebExtensionAPIScripting.InsertAndRemoveCSS "
+            isEnabled = "YES">
+         </CommandLineArgument>
+      </CommandLineArguments>
```
Revert.

kiaraarose replied:
oops! will do

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIScripting.mm:
```diff
@@ -31,14 +31,15 @@
     @"version": @"1",
 
-    @"permissions": @[ @"scripting" ],
+    @"permissions": @[ @"scripting", @"webNavigation" ],
```
Not using `webNavigation` currently.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIScripting.mm:
```diff
@@ -88,9 +89,277 @@
+
+    auto *urlRequest = server.requestWithLocalhost();
+    auto *url = urlRequest.URL;
+    auto *matchPattern = [_WKWebExtensionMatchPattern matchPatternWithScheme:url.scheme host:url.host path:@"/*"];
+    [manager.get().context setPermissionStatus:_WKWebExtensionContextPermissionStatusGrantedExplicitly forMatchPattern:matchPattern];
```
Can just do:

```
[manager.get().context setPermissionStatus:_WKWebExtensionContextPermissionStatusGrantedExplicitly forURL:urlRequest.URL];
```

It will make the pattern for you.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIScripting.mm:
```diff
@@ -88,9 +89,277 @@
+        @"background.js": backgroundScript,
+        @"backgroundColor.css": backgroundColor,
+        @"fontSize.css": fontSize,
+    }]);
```
Break `resources` out into a var.

kiaraarose committed:
Add remaining support for the browser.scripting API.
https://bugs.webkit.org/show_bug.cgi?id=259956

Reviewed by Timothy Hatcher.

This patch moves the code used for tabs.executeScript into the WebExtensionDynamicScripts so that
the code can be shared with the scripting.executeScript API.

Also adds tests for executeScript(), insertCSS(), and removeCSS().

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
(WebKit::WebExtensionContext::scriptingExecuteScript):
(WebKit::WebExtensionContext::scriptingInsertCSS):
(WebKit::WebExtensionContext::scriptingRemoveCSS):
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPITabsCocoa.mm:
(WebKit::WebExtensionContext::tabsExecuteScript):
(WebKit::WebExtensionContext::tabsInsertCSS):
(WebKit::WebExtensionContext::tabsRemoveCSS):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
(WebKit::WebExtensionDynamicScripts::getSourcePairsForParameters):
(WebKit::WebExtensionDynamicScripts::executeScript):
(WebKit::WebExtensionDynamicScripts::injectStyleSheets):
(WebKit::WebExtensionDynamicScripts::removeStyleSheets):
(WebKit::WebExtensionDynamicScripts::getSourcePairsForResource): Deleted.
* Source/WebKit/UIProcess/Extensions/WebExtensionDynamicScripts.h:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
(WebKit::parseTargetInjectionOptions):
(WebKit::WebExtensionAPIScripting::executeScript):
(WebKit::WebExtensionAPIScripting::toWebAPI):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIScripting.h:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/xcshareddata/xcschemes/TestWebKitAPI.xcscheme:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIScripting.mm:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/270576@main
